File: exampleApp/ios-native-src/BlePeripheralModule.m
#import "BlePeripheralModule.h"

static NSString *const kServiceUUID = @"12345678-1234-1234-1234-123456789abc";
static NSString *const kCharUUID    = @"abcdefab-1234-1234-1234-abcdefabcdef";

@implementation BlePeripheralModule

RCT_EXPORT_MODULE(BlePeripheral);

+ (BOOL)requiresMainQueueSetup { return NO; }

// ------------------------------------------------------------------
// JS-callable methods
// ------------------------------------------------------------------

RCT_EXPORT_METHOD(startPeripheral:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  self.pendingResolve = resolve;
  self.pendingReject  = reject;
  self.waitingToStart = YES;

  if (!self.peripheralManager) {
    self.peripheralManager = [[CBPeripheralManager alloc] initWithDelegate:self queue:nil];
    // Will call peripheralManagerDidUpdateState: when ready
  } else if (self.peripheralManager.state == CBManagerStatePoweredOn) {
    [self setupAndAdvertise];
  } else {
    reject(@"BT_OFF", @"Bluetooth is not powered on", nil);
    self.pendingResolve = nil;
    self.pendingReject  = nil;
  }
}

RCT_EXPORT_METHOD(stopPeripheral:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  self.waitingToStart = NO;
  [self.peripheralManager stopAdvertising];
  [self.peripheralManager removeAllServices];
  resolve(nil);
}

// ------------------------------------------------------------------
// CBPeripheralManagerDelegate
// ------------------------------------------------------------------

- (void)peripheralManagerDidUpdateState:(CBPeripheralManager *)peripheral {
  if (peripheral.state == CBManagerStatePoweredOn && self.waitingToStart) {
    [self setupAndAdvertise];
  } else if (peripheral.state != CBManagerStatePoweredOn) {
    if (self.pendingReject) {
      self.pendingReject(@"BT_OFF", @"Bluetooth is not powered on", nil);
      self.pendingResolve = nil;
      self.pendingReject  = nil;
    }
  }
}

- (void)setupAndAdvertise {
  [self.peripheralManager removeAllServices];

  self.pingCharacteristic = [[CBMutableCharacteristic alloc]
    initWithType:[CBUUID UUIDWithString:kCharUUID]
    properties:CBCharacteristicPropertyWrite | CBCharacteristicPropertyWriteWithoutResponse
    value:nil
    permissions:CBAttributePermissionsWriteable];

  CBMutableService *service = [[CBMutableService alloc]
    initWithType:[CBUUID UUIDWithString:kServiceUUID]
    primary:YES];
  service.characteristics = @[self.pingCharacteristic];

  [self.peripheralManager addService:service]; // → didAddService:error:
}

- (void)peripheralManager:(CBPeripheralManager *)peripheral
           didAddService:(CBService *)service
                   error:(NSError *)error {
  if (error) {
    if (self.pendingReject) {
      self.pendingReject(@"ADD_SERVICE_FAILED", error.localizedDescription, error);
      self.pendingResolve = nil;
      self.pendingReject  = nil;
    }
    return;
  }
  [peripheral startAdvertising:@{
    CBAdvertisementDataServiceUUIDsKey: @[service.UUID],
    CBAdvertisementDataLocalNameKey: @"PingApp"
  }]; // → peripheralManagerDidStartAdvertising:error:
}

- (void)peripheralManagerDidStartAdvertising:(CBPeripheralManager *)peripheral
                                       error:(NSError *)error {
  if (error) {
    if (self.pendingReject) {
      self.pendingReject(@"ADVERTISE_FAILED", error.localizedDescription, error);
    }
  } else {
    if (self.pendingResolve) {
      self.pendingResolve(nil);
    }
  }
  self.pendingResolve = nil;
  self.pendingReject  = nil;
}

// Respond to write-with-response requests; write-without-response needs no reply.
- (void)peripheralManager:(CBPeripheralManager *)peripheral
  didReceiveWriteRequests:(NSArray<CBATTRequest *> *)requests {
  for (CBATTRequest *req in requests) {
    [peripheral respondToRequest:req withResult:CBATTErrorSuccess];
  }
}

@end

File: exampleApp/modules/ble-peripheral/ios/BlePeripheralModule.swift
import ExpoModulesCore
import CoreBluetooth

private let kServiceUUID = CBUUID(string: "12345678-1234-1234-1234-123456789abc")
private let kCharUUID    = CBUUID(string: "abcdefab-1234-1234-1234-abcdefabcdef")

// ─── Expo Module ──────────────────────────────────────────────────────────────
// AsyncFunction handles (String, Promise) natively — no ObjC interop bug.

public class BlePeripheralModule: Module {
    private let ble = BlePeripheralDelegate()

    public func definition() -> ModuleDefinition {
        Name("BlePeripheral")

        // Atomic: message and start in one call
        AsyncFunction("startPeripheral") { (message: String, promise: Promise) in
            self.ble.start(message: message, promise: promise)
        }

        AsyncFunction("stopPeripheral") { (promise: Promise) in
            self.ble.stop()
            promise.resolve()
        }
    }
}

// ─── CBPeripheralManager delegate (separate class — needs NSObject) ───────────

private class BlePeripheralDelegate: NSObject, CBPeripheralManagerDelegate {
    private var manager:       CBPeripheralManager?
    private var pendingPromise: Promise?
    private var pendingMessage = ""

    func start(message: String, promise: Promise) {
        pendingMessage = message
        pendingPromise = promise

        if manager == nil {
            manager = CBPeripheralManager(delegate: self, queue: nil)
            // peripheralManagerDidUpdateState fires when ready
        } else if manager?.state == .poweredOn {
            setup()
        } else {
            promise.reject("BT_OFF", "Bluetooth is not powered on")
            pendingPromise = nil
        }
    }

    func stop() {
        manager?.stopAdvertising()
        manager?.removeAllServices()
        pendingMessage = ""
    }

    // ── State machine ─────────────────────────────────────────────────────────

    private func setup() {
        manager?.removeAllServices()

        // Static value — CoreBluetooth answers reads automatically, no delegate needed
        let char = CBMutableCharacteristic(
            type: kCharUUID,
            properties: .read,
            value: pendingMessage.data(using: .utf8),
            permissions: .readable
        )
        let service = CBMutableService(type: kServiceUUID, primary: true)
        service.characteristics = [char]
        manager?.add(service) // → peripheralManager(_:didAdd:error:)
    }

    func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        if peripheral.state == .poweredOn, pendingPromise != nil {
            setup()
        } else if peripheral.state != .poweredOn {
            pendingPromise?.reject("BT_OFF", "Bluetooth is not powered on")
            pendingPromise = nil
        }
    }

    func peripheralManager(_ peripheral: CBPeripheralManager,
                           didAdd service: CBService, error: Error?) {
        if let error = error {
            pendingPromise?.reject("ADD_SERVICE_FAILED", error.localizedDescription)
            pendingPromise = nil
            return
        }
        peripheral.startAdvertising([
            CBAdvertisementDataServiceUUIDsKey: [service.uuid],
            CBAdvertisementDataLocalNameKey: "MsgApp",
        ]) // → peripheralManagerDidStartAdvertising(_:error:)
    }

    func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager,
                                              error: Error?) {
        if let error = error {
            pendingPromise?.reject("ADVERTISE_FAILED", error.localizedDescription)
        } else {
            pendingPromise?.resolve()
        }
        pendingPromise = nil
    }
}
